File: frontend/src/lib/materials.ts
export type ShaderType = "toon" | "pbr" | "metallic" | "holographic" | "matte" | "emissive";

export interface MaterialProfile {
  id: string;
  name: string;
  shader_type: ShaderType;
  base_color: string;
  secondary_color: string;
  emissive_color: string;
  metallic: number;
  roughness: number;
  emissive_intensity: number;
  opacity: number;
  rim_light: boolean;
  toon_steps: number;
  is_preset: boolean;
  created_at: string;
  updated_at: string;
}

export const SHADER_LABELS: Record<ShaderType, string> = {
  toon: "Toon",
  pbr: "PBR",
  metallic: "Metallic",
  holographic: "Holographic",
  matte: "Matte",
  emissive: "Emissive",
};

export const SHADER_TYPES: ShaderType[] = ["toon", "pbr", "metallic", "holographic", "matte", "emissive"];

export async function listMaterials(): Promise<MaterialProfile[]> {
  const res = await fetch("/api/materials");
  if (!res.ok) throw new Error("Failed to fetch materials");
  return res.json();
}

export async function createMaterial(data: Partial<MaterialProfile>): Promise<MaterialProfile> {
  const res = await fetch("/api/materials/create", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail ?? "Failed to create material");
  }
  return res.json();
}

export async function updateMaterial(id: string, data: Partial<MaterialProfile>): Promise<MaterialProfile> {
  const res = await fetch(`/api/materials/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail ?? "Failed to update material");
  }
  return res.json();
}

export async function deleteMaterial(id: string): Promise<void> {
  const res = await fetch(`/api/materials/${id}`, { method: "DELETE" });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error((err as { detail?: string }).detail ?? "Failed to delete material");
  }
}

export async function activateMaterial(id: string): Promise<MaterialProfile> {
  const res = await fetch(`/api/materials/${id}/activate`, { method: "POST" });
  if (!res.ok) throw new Error("Failed to activate material");
  return res.json();
}

File: backend/app/routes/materials.py
from __future__ import annotations
from fastapi import APIRouter
from pydantic import BaseModel

from app.models.materials import MaterialProfile
from app.services import material_service

router = APIRouter(prefix="/api/materials", tags=["materials"])


class MaterialCreateRequest(BaseModel):
    name: str = "Custom Material"
    shader_type: str = "pbr"
    base_color: str = "#1e293b"
    secondary_color: str = "#334155"
    emissive_color: str = "#06b6d4"
    metallic: float = 0.5
    roughness: float = 0.5
    emissive_intensity: float = 0.2
    opacity: float = 1.0
    rim_light: bool = False
    toon_steps: int = 3


class MaterialUpdateRequest(BaseModel):
    name: str | None = None
    shader_type: str | None = None
    base_color: str | None = None
    secondary_color: str | None = None
    emissive_color: str | None = None
    metallic: float | None = None
    roughness: float | None = None
    emissive_intensity: float | None = None
    opacity: float | None = None
    rim_light: bool | None = None
    toon_steps: int | None = None


@router.get("", response_model=list[MaterialProfile])
async def list_materials():
    return material_service.list_materials()


@router.post("/create", response_model=MaterialProfile, status_code=201)
async def create_material(body: MaterialCreateRequest):
    return material_service.create_material(body.model_dump(exclude_none=False))


@router.patch("/{material_id}", response_model=MaterialProfile)
async def update_material(material_id: str, body: MaterialUpdateRequest):
    return material_service.update_material(
        material_id, body.model_dump(exclude_none=True)
    )


@router.delete("/{material_id}", status_code=204)
async def delete_material(material_id: str):
    material_service.delete_material(material_id)


@router.post("/{material_id}/activate", response_model=MaterialProfile)
async def activate_material(material_id: str):
    return material_service.activate_material(material_id)
